python/xir/program.py

```diff
@@ -201,19 +203,31 @@ class Declaration:
 
     Args:
         name (str): name of the declaration
+        declaration_type (str): The type of declaration. Can be either "gate", "operator", "output"
+            or "function".
+        params (Sequence[str]): parameters used by the declared object
+        wires (Sequence[Wire]): wires that the declared object is applied to
     """
 
     def __init__(
-        self, name: str, params: Sequence[str], wires: Sequence[Wire], declaration_type: str
+        self,
+        name: str,
+        declaration_type: str,
+        params: Optional[Sequence[str]] = None,
+        wires: Optional[Sequence[Wire]] = None,
     ) -> None:
-        self.name = name
-        self.params = list(map(str, params))
-        if len(set(self.params)) != len(self.params):
-            raise ValueError("All parameters in a declaration must be unique.")
-        self.wires = wires
         if declaration_type not in ("gate", "output", "operator", "function"):
             raise TypeError(f"Declaration type '{declaration_type}' is invalid.")
+
+        self.name = name
         self.declaration_type = declaration_type
+        self.params = params or []
+        self.wires = wires or ()
+
+        if not all(isinstance(p, str) for p in self.params):
+            raise TypeError("Declared parameters must be strings.")
+        if len(set(self.params)) != len(self.params):
+            raise ValueError("All parameters in a declaration must be unique.")
```

Comment:
Although these error messages are concise, a user may not find them helpful without some additional context.  What do you think about putting the parameters (or the name of the declaration) in the message?
```suggestion
        if not all(isinstance(p, str) for p in self.params):
            raise TypeError("Declaration '{name}' has parameters which are not strings.")
        if len(set(self.params)) != len(self.params):
            raise ValueError("Declaration '{name}' has duplicate parameters.")
```